File: fsharp/gigasecond/Gigasecond.fs
﻿//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//                     Version 2, December 2004
//
//  Copyright (C) 2016 Eric Bailey <eric@ericb.me>
//
//  Everyone is permitted to copy and distribute verbatim or modified
//  copies of this license document, and changing it is allowed as long
//  as the name is changed.
//
//             DO WHAT THE FUCK YOU WANT TO PUBLIC LICENSE
//    TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION
//
//   0. You just DO WHAT THE FUCK YOU WANT TO.
//

module Gigasecond

open System

/// Given a `DateTime`, return the `Date` a gigasecond later.
let gigasecond (input : DateTime) : DateTime = input.AddSeconds(1e9).Date

File: erlang/gigasecond/gigasecond.erl
%% =============================================================================
%% @author Eric Bailey
%% @copyright 2015 Eric Bailey
%% @version 0.1.0
%% @doc Calculating gigasecond anniversaries.
%% @end
%% =============================================================================

-module(gigasecond).

-export([from/1]).

%% A gigasecond is one billion (10**9) seconds.
-define(GIGASECOND, 1000000000).

%% @doc Given a `calendar:date()' or a `calendar:datetime()', return a
%% `calendar:datetime()' a gigasecond later.
-spec from(calendar:date() | calendar:datetime()) -> calendar:datetime().
from(Date = {_Y, _M, _D}) ->
    from({Date, {0, 0, 0}});
from(DateTime) ->
    Seconds = calendar:datetime_to_gregorian_seconds(DateTime),
    calendar:gregorian_seconds_to_datetime(Seconds + ?GIGASECOND).
